File: src/csharp/Foundgine.Core/Semantic/Planning/SemanticPlanOptimizationResult.cs
namespace Foundgine.Core.Semantic.Planning;

/// <summary>Result of provider-neutral semantic plan optimization.</summary>
public sealed record SemanticPlanOptimizationResult(
    SemanticPlan Plan,
    IReadOnlyList<string> AppliedRules,
    SecurityPreservationProof SecurityProof,
    SemanticEquivalenceProof SemanticProof,
    SemanticPlanAuthorizationBindingProof AuthorizationBindingProof,
    IReadOnlyList<PlanRewriteRuleResult>? RuleApplications = null,
    double TotalCostImpact = 0d,
    bool TerminatedNormally = true)
{
    public bool Changed => AppliedRules.Count != 0;
    public IReadOnlyList<PlanRewriteRuleResult> Applications => RuleApplications ?? [];
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/SemanticPlanOptimizationResult.java
package com.foundgine.core.semantic.planning;

import java.util.*;

/** Result of provider-neutral semantic plan optimization. */
public record SemanticPlanOptimizationResult(SemanticPlan plan, List<String> appliedRules,
		SecurityPreservationProof securityProof, SemanticEquivalenceProof semanticProof,
		SemanticPlanAuthorizationBindingProof authorizationBindingProof, List<PlanRewriteRuleResult> ruleApplications,
		double totalCostImpact, boolean terminatedNormally) {
	public SemanticPlanOptimizationResult {
		appliedRules = List.copyOf(appliedRules);
		ruleApplications = ruleApplications == null ? List.of() : List.copyOf(ruleApplications);
	}

	public boolean changed() {
		return !appliedRules.isEmpty();
	}

	public List<PlanRewriteRuleResult> applications() {
		return ruleApplications;
	}
}
